### DISCUSSION TITLE
how can i install jdk? (on railway)

### DISCUSSION THREAD

**Discussion by wasaji** [coder/code-server#4109] [Q&A]
I cannot use java extensions because I do not have jdk

**Comment by lg-nineoneninethreenine** [comment]
What is JDK? I have never heard of it that's why.

**Reply by wasaji** [reply]
it means Java Development Kit

**Reply by lg-nineoneninethreenine** [reply]
So, typically, you need Java, right?

**Reply by wasaji** [reply]
yes that's it

**Reply by lg-nineoneninethreenine** [reply]
Okay, be right back.

**Reply by wasaji** [reply]
Thx, keep waiting for you dude

**Reply by lg-nineoneninethreenine** [reply]
I think this is the right guide for getting Java on Linux(yes, the Railway terminal is Debian Buster), see [here](https://tecadmin.net/how-to-install-java-on-debian-10-buster/)

**Reply by wasaji** [reply]
How can i use railway terminal? I don't know where can I use that.

**Reply by lg-nineoneninethreenine** [reply]
Hit that hamburger button on the top-left side and hover over the Terminal section(click it if you want). Then, you get a lot of options(5-6 options estimate, idk forgot that). Hit the first one. Then you get a Bash terminal and you are user `coder`.

**Reply by lg-nineoneninethreenine** [reply]
You shouldn't change that because if you want the base user coder and you made and switched to another user, it's gonna be hard(very hard)

**Reply by wasaji** [reply]
hamburger button?? I cannot understand what you say. cuz I am not goot at English.. and when I click that railway icon on the top-left side, it just goes to "https://railway.app/dashboard". I cannot find terminal section

**Reply by lg-nineoneninethreenine** [reply]
Wait, I will be back.

**Reply by lg-nineoneninethreenine** [reply]
Step 1: Hit this button:
![image](https://user-images.githubusercontent.com/76845820/132174750-1469ddfa-885a-4b95-af46-ad6467742173.png)
Step 2: Hit the Terminal section:
![image](https://user-images.githubusercontent.com/76845820/132174949-48f7206b-ae6f-48b4-9877-0ab48c8150f7.png)
Step 3: Hit the `New Terminal` Button:
![image](https://user-images.githubusercontent.com/76845820/132175111-89c370a2-ac80-4aa7-86dd-85e67c465405.png)

You will have a terminal that looks (almost) similar to this:
![image](https://user-images.githubusercontent.com/76845820/132175434-9f39c594-ddd8-4657-8b08-ea78ea730868.png)

Note: I use the `sudo su` command so I can switch **temporarily** to the `im-coder-lg` account. You can also do that but DO NOT, under any circumstances, remove the base `coder` account.

**Reply by wasaji** [reply]
So, I've just open new terminal. then can I type "sudo apt install openjdk-8-jdk" on the terminal?

**Reply by wasaji** [reply]
![image](https://user-images.githubusercontent.com/86885316/132179034-53e0b4d3-7f4f-44d8-ba84-a7b295eba8ce.png)
unable to install..

**Reply by lg-nineoneninethreenine** [reply]
idk dude it is confusing. Try searching `how to install java development kit on debian` and check if any of them work.

**Reply by wasaji** [reply]
Once I installed jdk8, but code-server didn't identified it.. It seems that Environment Variable is not set. What should I do?

**Reply by lg-nineoneninethreenine** [reply]
@wasaji idk only @jsjoeio can tell. You will have to wait for the reply.

**Reply by wasaji** [reply]
OK dude. Thank you so much for your heartful replies

**Reply by lg-nineoneninethreenine** [reply]
Anytime. But if you need more deep help for JDK, one of the maintainers will get in touch soon.

**Comment by lg-nineoneninethreenine** [comment]
@Wasaji look at this link: https://docs.oracle.com/en/java/javase/11/install/installation-jdk-linux-platforms.html#GUID-4A6BD592-1840-4BB4-A758-4CD49E9EE88B
Now that you know how to open the Terminal, type this:
```bash
sudo apt update
sudo apt install neofetch
neofetch
```
If you see Neofetch, it tells that it runs on an Intel Xeon(for me) but might vary for you. I will tell how to download and unzip the tar archive later.

**Reply by wasaji** [reply]
tar archive? idk what it is lol.
![image](https://user-images.githubusercontent.com/86885316/132225908-e9f66495-2ea9-4a8d-b0b1-8a410ca6bcd9.png)
I can see this when I type neofetch

**Reply by lg-nineoneninethreenine** [reply]
Yup, it is x64... Posting the guide soon and a tar archive is like the official archive of Linux. Windows has zip as default archive format, macOS has something like that. Linux has a lot of archive formats. Tar is one of them.

**Reply by wasaji** [reply]
Ok. but umm.. I've already installed jdk11 and jdk8 I think I cannot run java code because I cannot set the environment variable

**Reply by lg-nineoneninethreenine** [reply]
You can set environment variables in Railway's Variables section. Your Git repo and your password is there by default but you can add new variables.

**Reply by wasaji** [reply]
![image](https://user-images.githubusercontent.com/86885316/132232925-b5c3e1c9-e77e-48d0-8bf7-1472c3d3f8f3.png)
I've reached here but idk what to type to set environment variables

**Reply by lg-nineoneninethreenine** [reply]
You have to name the env var(environment variables) as you name it when you install JDK, I guess.

**Reply by wasaji** [reply]
I cannot understand what you say..

**Reply by lg-nineoneninethreenine** [reply]
What is your environment variable? It has a name and a value, right? In that `NEW_VARIABLE` box, add your variable name in caps lock(that key on your keyboard) and use underscores(`_`) for spaces. Then, in the `VALU3` box, add your value and hit the `Add` button and try again(wait for deployment publish)

**Reply by wasaji** [reply]
I don't know what env var i actually.
![image](https://user-images.githubusercontent.com/86885316/132301244-224707be-7a5a-45e3-8428-ed025c2607b4.png)
What should I do?

**Reply by lg-nineoneninethreenine** [reply]
idk only an experienced person can tell.

**Reply by jsjoeio** [reply]
I am not a Java developer so I don't think I can help much. It seems like you have Java installed in your Railway environment.

I'm not sure what the values of `JDK_HOME` and `JAVA_HOME` should be 🤔 

Here's how I would try to figure it out:
1. `JAVA_HOME` needs to point to the location where Java is installed in your environment: see [here](https://stackoverflow.com/a/4681109/3015595). Once you know that, you can set that value in Railway's environment variables
2. Do the same for `JDK_HOME`

Something else that might be helpful: https://www.linuxfordevices.com/tutorials/linux/set-the-java-home-variable

**Reply by wasaji** [reply]
thx but  I cannot try that cuz my account is banned
![image](https://user-images.githubusercontent.com/86885316/132427383-3582b728-d4ab-4315-985f-38f3bc2508e0.png)
wtf is this?? LOL
Is it better to make a new account rather than sending an email?

**Reply by lg-nineoneninethreenine** [reply]
I think this might help.
https://itsfoss.com/run-java-program-ubuntu/

**Reply by lg-nineoneninethreenine** [reply]
And create a new account via mail.

**Reply by jsjoeio** [reply]
"Crytpo miner" lol uh...

Yeah might be worth reaching out to Railway support then 😅 I'm not sure on that one

**Reply by lg-nineoneninethreenine** [reply]
@jsjoeio looks like the environment variables made Railway think that @Wasaji was a cryptocurrency miner, looks like the variables made that.

**Reply by wasaji** [reply]
@jsjoeio @im-coder-lg 
![image](https://user-images.githubusercontent.com/86885316/132733476-54cc1632-4f3e-4ed0-b2e4-3056a57b5f6b.png)
I cannot send an email cuz my chrome tab becomes like this whenever I enter the railway site..
What should I do?

**Reply by jsjoeio** [reply]
Can you try reaching out to them on [Twitter](https://twitter.com/Railway)?

**Reply by lg-nineoneninethreenine** [reply]
Actually @Wasaji had to add the JAVA_HOME variable inside bash like this: https://itsfoss.com/set-java-home-ubuntu/
The thing is it's Ubuntu but could work on Debian. If @wasaji obtained a Railway account via email or any other methods, he can get code-server deployed and test it this way.

**Reply by wasaji** [reply]
I've just got a new account via a new github acc but railway blocked me LOL
![image](https://user-images.githubusercontent.com/86885316/133827008-521d0ef0-245d-4110-a57b-aca0d8c95f59.png)

**Reply by wasaji** [reply]
@jsjoeio I don't know what their email is. Can you tell me what it is?

**Reply by jsjoeio** [reply]
Oh no! I don't know it either, but they have a Discord you can join and get help? [link](https://discord.com/invite/xAm2w6g)

**Reply by wasaji** [reply]
I've just entered the server dude

**Reply by jsjoeio** [reply]
Nice dude! Well hoping you can get some help 🤞

**Reply by lg-nineoneninethreenine** [reply]
Actually, in that page, if you hit the `Email` word, you can mail Railway Support.

**Reply by wasaji** [reply]
@jsjoeio @im-coder-lg I was finally unbanned and now I know how to install jdk using dockerfile however the debugger does not work what should I do?

**Reply by jsjoeio** [reply]
I think this is an issue with all the debuggers in VS Code. Subscribe to [this issue](https://github.com/cdr/code-server/issues/4221). Hoping to fix soon!
